### ISSUE TITLE
image datatypes

### ISSUE THREAD

**Issue by d-v-b** [napari/napari#1013]
rgba data in numpy is usually represented as an array with n spatial dimensions and 1 channel dimension that has length 3 (for rgb) or length 4 (for rgba). This is a little unfortunate, because the rgba-ness of the data is a property of the values, not of the array of values. This leads to stuff in napari like specifying which array dimensions correspond to the channel (no slider) vs the array dimensions that are really dimensions (slider).

In numpy, you can define custom datatypes for elements of a numpy array; the docs actually [demonstrate](https://docs.scipy.org/doc/numpy/reference/arrays.dtypes.html) using this feature for generating a datatype that represents a single rgba sample. With this datatype, you can then produce an array of rgba values which doesn't have a pseudo-dimension for the color channels. Practically, it ~~should be~~ is possible to use the `np.view` to generate a view of a ND + C array as an ND array of C-length values:

```python
dt = np.dtype(('i4', [('r', 'u1'), ('g', 'u1'), ('b', 'u1'), ('a', 'u1')]))
tmp = np.random.randint(0, 255, (10,10,4), dtype='uint8')
v = tmp.view(dt)[0,0] # array([658229181], dtype=(numpy.int32, [('r', 'u1'), ('g', 'u1'), ('b', 'u1'), ('a', 'u1')]))
v['r'] #  array([189], dtype=uint8)
v.view('uint8') # array([189, 199,  59,  39], dtype=uint8)
```

Within napari, using `dtype` machinery for `rgba` (and maybe more channels, for hyperspectral data) would be really convenient for separating array properties (the layout of the values) from value properties (the stuff you get from accessing a single element of the array).

But this is old functionality in numpy, and there's probably a good reason why (to my knowledge) libraries like skimage exclusively use a specific array axis for representing the color channel. @jni has this idea come up before?

**Comment by jni** [comment]
Interesting idea! No, I haven't seen this use of dtypes or recarrays brought up before. Having said this, though, this is non-standard, and I am reluctant to have us make a push for making it standard, as I think it has limited value compared to xarray, which, as you point out, we will probably end up adopting at some point, and then this problem could be solved equally well with that. In general, I think recarrays and custom dtypes are not used very often with NumPy and this means that people will be surprised to see such code, and also (more speculative) I suspect that within NumPy their performance will be slightly less optimised because of this.

In fact, this approach prevents you from (or makes it harder) reordering the axes to optimise cache access for particular algorithms (e.g. denoising channels independently of each other), or doing linear algebra or convolutions on the channel axis, for example.

So, imho we should leave this alone as an "interesting mathematical equivalence" rather than attempt to make it our next workhorse.

**Comment by GenevieveBuckley** [comment]
I'm +1 on xarray instead of custom dtypes as well.

**Comment by d-v-b** [comment]
Yeah I agree that doing strange stuff should be avoided when possible... But, if this datatype were used internally, I don't think it would actually be that weird: every user with C-ordered array data in the form X,Y,...C is already compliant with the scheme I depicted above (since i'm proposing a view of data, not a transformation of data). Both matplotlib and vispy assume rgb numpy arrays will come in X,Y,C format. So, at visualization time, this convention is already forced on users by the existing visualization tools.

Also, I don't think this is an "interesting mathematical equivalence" -- I view it more of formally admitting the truth: _color is not a real dimension_ in rgba images. Color is a property of the _elements_ of the array. I am prepared to die on this hill :)

As I think more about implementing an rgba datatype, it wouldn't obviously solve any problems -- `guess_rgb` and `is_rgb` might get moved around but that logic would still be needed. So until this solves real problems I'll just keep it in the back of my mind.

**Comment by jni** [comment]
@d-v-b note that in scikit-image, we will no longer assume the last axis is the channel axis. See https://github.com/scikit-image/scikit-image/issues/4294. And this is very important in microscopy where it is *often* not the last axis. Due to acquisition order, many microscopy arrays come in TZCYX order.

At any rate, I don't like having lots of code that assumes RGB, which a tiny sample in an enormous universe of multispectral data. I don't think NumPy dtype is the right model for this, vs axis annotations. [HyperSpy](http://hyperspy.org/hyperspy-doc/current/user_guide/getting_started.html) uses the idea of "signal axes" (roughly: channels) and "navigation axes" (roughly: spatio-temporal axes). Eric Prestat introduced me to the idea of an "image" where at every "pixel" there is a 2D X-ray diffraction pattern. I think the dtype approach will start to creak rather quickly in scenarios like this one.

And again, when it comes time to processing the spectra, you *have* to think of them as an array of values, e.g. http://hyperspy.org/hyperspy-doc/current/user_guide/eds.html#geting-the-intensity-of-an-x-ray-line .

**Comment by d-v-b** [comment]
@jni good points! Other sample datatypes to consider are vectors (for use in vector fields), and LabelledMultisets, which get used in segmentation images for reversibly aggregating labelled samples. The latter are not fixed-length, which complicates the "just add another array axis" approach.  

> Eric Prestat introduced me to the idea of an "image" where at every "pixel" there is a 2D X-ray diffraction pattern. I think the dtype approach will start to creak rather quickly in scenarios like this one.

[Image scanning microscopy](https://physics.aps.org/featured-article-pdf/10.1103/PhysRevLett.104.198101), a generalization of the confocal microscopy, also generates raw data like this.

I will keep thinking about this issue. Although I think axis annotations are a great idea in general, I think they solve a different problem.
